demos/common/processing/async_pipeline.h

```diff
@@ -0,0 +1,150 @@
+
+        double getLastInferenceLatencyMs() const {
+            return ((double)std::chrono::duration_cast<std::chrono::milliseconds>(lastInferenceLatency).count());
+        }
+
+    };
+
+    struct PerformanceInternalCounters
+    {
+        long long latenciesMs[MOVING_AVERAGE_SAMPLES] = {};
+        std::chrono::steady_clock::time_point retrievalTimestamps[MOVING_AVERAGE_SAMPLES] = {};
+        long long movingLatenciesSumMs = 0;
+        int currentIndex = 0;
+    };
+
+public:
+    /// Loads model and performs required initialization
+    /// @param modelInstance pointer to model object. Object it points to should not be destroyed manually after passing pointer to this function.
+    /// @param cnnConfig - fine tuning configuration for CNN model
+    /// @param engine - pointer to InferenceEngine::Core instance to use.
+    /// If it is omitted, new instance of InferenceEngine::Core will be created inside.
+    PipelineBase(ModelBase* modelInstance, const CnnConfig& cnnConfig, InferenceEngine::Core* engine = nullptr);
+    virtual ~PipelineBase();
+
+    /// Waits until either output data becomes available or pipeline allows to submit more input data.
+    /// @param shouldKeepOrder if true, function will treat results as ready only if next sequential result (frame) is
+    /// ready (so results can be extracted in the same order as they were submitted). Otherwise, function will return if any result is ready.
+    void waitForData(bool shouldKeepOrder = true);
+
+    /// Returns true if there's available infer requests in the pool
+    /// and next frame can be submitted for processing.
+    /// @returns true if there's available infer requests in the pool
+    /// and next frame can be submitted for processing, false otherwise.
+    bool isReadyToProcess() { return requestsPool->isIdleRequestAvailable(); }
+
+    /// Returns performance info
+    /// @returns performance information structure
+    PerformanceInfo getPerformanceInfo() { std::lock_guard<std::mutex> lock(mtx); return perfInfo; }
+
+    /// Waits for all currently submitted requests to be completed.
+    ///
+    void waitForTotalCompletion() { if(requestsPool.get())requestsPool->waitForTotalCompletion(); }
+
+    /// Submit request to network
+    /// @param image - image to submit for processing
+    /// @returns -1 if image cannot be scheduled for processing (there's no any free InferRequest available).
+    /// Otherwise reqturns unique sequential frame ID for this particular request. Same frame ID will be written in the responce structure.
+    virtual int64_t submitImage(cv::Mat img);
+
+    /// Gets available data from the queue 
+    /// @param shouldKeepOrder if true, function will treat results as ready only if next sequential result (frame) is
+    /// ready (so results can be extracted in the same order as they were submitted). Otherwise, function will return if any result is ready.
+    virtual std::unique_ptr<ResultBase> getResult(bool shouldKeepOrder=true);
```

Comment:
`PipelineBase` seems to try to be general enough to be used with other instances of `PipelineBase` of different models. We didn't work out a good idea how to write such pipelines and thus I have a concern that we can design such class without having that idea in mind.
In particular there can be a situation when we need to extract `ResultBase` which is ready and has the smallest `frameId`, while this function can only provide `ResultBase` with "random" `frameId` or with next `frameId`. In this case the solution is to replace unordered_map with map, but the problem is that there can be other features to consider.

My suggestion is to focus on single model pipeline and not to introduce useless from that point of view features.

Reply:
"Too general" sounds too general. But, as we discussed, I've removed shouldKeepOrder.